runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -325,83 +326,103 @@ protected void initChannel(@Nonnull Channel ch) throws Exception {
-                    }
+                log.info("addReconnectionOnCloseFuture[{}]: reconnecting", node);
+                // Asynchronously connect again.
+                connectAsync(b);
+            }
+        });
+    }
+
+    /** Connect to a remote server asynchronously.
+     *
+     * @param b         The channel boostrap to use
+     * @return          A {@link ChannelFuture} which is c
+     */
+    private ChannelFuture connectAsync(@Nonnull Bootstrap b) {
+        // If shutdown, return a ChannelFuture that is exceptionally completed.
+        if (shutdown) {
+            return new DefaultChannelPromise(channel, GlobalEventExecutor.INSTANCE)
+                .setFailure(new ShutdownException("Runtime already shutdown!"));
+        }
+        // Use the bootstrap to create a new channel.
+        ChannelFuture f = b.connect(node.getHost(), node.getPort());
+        f.addListener((ChannelFuture future) -> {
+            if (future.isSuccess()) {
+                // When the channel connection succeeds, set this channel as active
+                channel = future.channel();
+                // And register a future to reconnect in case we get disconnected
+                addReconnectionOnCloseFuture(channel, b);
+                log.info("connectAsync[{}]: Channel connected.", node);
+            } else {
+                // Otherwise, the connection failed. If we're not shutdown, try reconnecting after
+                // a sleep period.
+                if (!shutdown) {
+                    log.info("connectAsync[{}]: Channel connection failed, reconnecting...", node);
+                    Sleep.sleepUninterruptibly(parameters.getConnectionRetryRate());
+                    // Call connect, which will retry the call again.
+                    connectAsync(b);
                 }
             }
         });
-        connected = true;
+        return f;
     }
 
     /**
      * Stops routing requests.
      */
     @Override
     public void stop() {
-        stop(false);
+        log.debug("stop: Shutting down router for {}", node);
+        shutdown = true;
+        connectionFuture.completeExceptionally(new ShutdownException());
+        try {
+            channel.disconnect();
+            channel.close().syncUninterruptibly();
+        } catch (Exception e) {
+            log.error("Error in closing channel");
+        }
+        try {
+            if (shutdownEventLoop) {
+                eventLoopGroup.shutdownGracefully().sync();
+            }
+        } catch (InterruptedException e) {
+            throw new UnrecoverableCorfuInterruptedError("Interrupted while stopping", e);
+        }
     }
 
+    /** {@inheritDoc}
+     *  @deprecated  Deprecated, stopping a router without shutting it down is no longer supported.
+     *               Please use {@link this#stop()}.
+     */
     @Override
+    @Deprecated
     public void stop(boolean shutdown) {
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Do not forget to remove this deprecated code someday. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)